mypy/sametypes.py

```diff
@@ -125,7 +125,7 @@ def visit_typeddict_type(self, left: TypedDictType) -> bool:
     def visit_union_type(self, left: UnionType) -> bool:
         # XXX This is a test for syntactic equality, not equivalence
         if isinstance(self.right, UnionType):
-            return is_same_types(left.items, self.right.items)
+            return is_same_types(sorted(left.items, key=str), sorted(self.right.items, key=str))
```

Comment:
I suspect that `str` doesn't always work, since somes types have internal state that doesn't affect equality but is visible in the string representation. A better approach would be to just bite the bullet and use quadratic, pairwise `is_same_type` comparisons.